### ISSUE TITLE
Experience Continuity - Two different visitors are bucketed if `userId` is loaded async

### ISSUE THREAD

**Issue by drazenbuljovcic** [flagship-io/flagship-react-sdk#107]
Type **Bug**

## Repro and explanation

We are having issues translating variation assignments from "two" different users - so the experience stays continuous.
 Generated `visitorId` does not get translated to `anonymousId` for an async "userId".

> Code used to initialize Flagship
```js
export const FlagshipProvider = ({ children }) => {
  const [visitorId, setVisitorId] = useState(null);

  useEffect(() => {
    internalAuthProvider().then(({ userId }) => {
      setVisitorId(userId);
    });
  }, []);

  return (
    <FlagshipSDKProvider
      fetchNow
      envId={ENV.FLAGSHIP_ENVIRONMENT_ID}
      apiKey={ENV.FLAGSHIP_API_KEY}
      visitorData={{
        id: visitorId, // let the sdk generate a temporary id
        context: {},
        isAuthenticated: !!visitorId,
        hasConsented: true,
        // link this visitor with our internal identifier
        // isAuthenticated: should be true after we call fs.authenticate
      }}
    >
      {children}
    </FlagshipSDKProvider>
  );
};
```

`visitorId` is set into state in an **async** way.

- The generated `uuid4` is done by the SDK at stage 1.
- The promise from our `internalAuthProvider` resolves and the `userId` is set into state.
- The SDK gets a new `visitorId` at stage 2.
- The next call to activate campaigns sends `anonymousId` -- another randomly generated `uuidv4` ❗  
- The "new" visitor is bucketed in different ways

## SDK calls to the decisionApi
**call1 to `campaigns` endpoint**
`visitorId: "tempUserId"`

**call2 to `campaigns` endpoint**
`visitorId: "abcdefg"` (not **tempUserId** ❓ )
`visitorId: "<internalId>"

https://user-images.githubusercontent.com/18490172/236245751-5d6479c4-8c50-41e3-aaa5-c6265995b930.mov

![236244134-50c3eb2f-67dd-4f9c-88c6-48564f4f3e93](https://user-images.githubusercontent.com/18490172/236247092-83ef249e-2881-4125-a1ce-180e135f2ad0.png)
![236244279-e68af9c8-6661-42d8-9bfb-759342c0c0af](https://user-images.githubusercontent.com/18490172/236247166-53c6956f-0eb5-428c-91aa-a0394280e8dc.png)

## Expected behavior
**vistorId** (step1) === **anonymousId** (step2)

Thank you!!

**Comment by Heihokon** [comment]
Hi @drazenbuljovcic 

Thank you for your report.

The fix is published in alpha version [3.1.3-alpha.0](https://www.npmjs.com/package/@flagship.io/react-sdk/v/3.1.3-alpha.0)

`yarn add @flagship.io/react-sdk@alpha`
`npm install @flagship.io/react-sdk@alpha`

You can therefore continue your integration while waiting for the latest release which will be released soon.

We'll let you know as soon as it's released.

**Comment by drazenbuljovcic** [comment]
Hi there @Heihokon !
Thank you so much for addressing this!

I can confirm that with the alpha version the `visitorId` from call1 is correctly translated to `anonymousId` in call2.

![image](https://user-images.githubusercontent.com/18490172/236426424-38fe3a3f-9099-4147-b13e-c66254d7ca6c.png)

```json
{
    "visitorId": "8dc30ce5-08b9-43ae-81cd-25579122e919",
    "campaigns": [
        {
            "id": "cgqk2qgel9a4tgrm7cpg",
            "slug": "yellow",
            "type": "ab",
            "variationGroupId": "cgqk2qgel9a4tgrm7cqg",
            "variation": {
                "id": "cgqk2qgel9a4tgrm7cr0",
                "modifications": {
                    "type": "FLAG",
                    "value": {
                        "container-background": "white"
                    }
                },
                "reference": true
            }
        },
        {
            "id": "cgqjo6c5vfp27i1h9qag",
            "slug": "red",
            "type": "ab",
            "variationGroupId": "cgqjo6c5vfp27i1h9qbg",
            "variation": {
                "id": "cgqjo6c5vfp27i1h9qc0",
                "modifications": {
                    "type": "FLAG",
                    "value": {
                        "edit-button-background": "whiteWithBorder"
                    }
                },
                "reference": true
            }
        }
    ]
}
```

![image](https://user-images.githubusercontent.com/18490172/236426746-868c3185-7874-4c74-a3c0-ba9422b02674.png)

```json
{
    "visitorId": "jogdh8idoyv6j5uu0k4tpgdazhcdtxnl",
    "campaigns": [
        {
            "id": "cgqk2qgel9a4tgrm7cpg",
            "slug": "yellow",
            "type": "ab",
            "variationGroupId": "cgqk2qgel9a4tgrm7cqg",
            "variation": {
                "id": "cgqk2qgel9a4tgrm7crg",
                "modifications": {
                    "type": "FLAG",
                    "value": {
                        "container-background": "#ffe783"
                    }
                },
                "reference": false
            }
        },
        {
            "id": "cgqjo6c5vfp27i1h9qag",
            "slug": "red",
            "type": "ab",
            "variationGroupId": "cgqjo6c5vfp27i1h9qbg",
            "variation": {
                "id": "cgqjo6c5vfp27i1h9qc0",
                "modifications": {
                    "type": "FLAG",
                    "value": {
                        "edit-button-background": "whiteWithBorder"
                    }
                },
                "reference": true
            }
        },
        {
            "id": "cgmnqq9rodhmhd4d19l0",
            "slug": "green",
            "type": "ab",
            "variationGroupId": "cgmnqq9rodhmhd4d19m0",
            "variation": {
                "id": "cgmnqq9rodhmhd4d19mg",
                "modifications": {
                    "type": "FLAG",
                    "value": {
                        "preview-button-background": "whiteWithBorder"
                    }
                },
                "reference": true
            }
        }
    ]
}
```

What I still see happening is that the visitor gets re-bucketed to different variations on change of the `visitorId`.
Is this expected behavior - or should we expect to see the same variations returned by the decision API in this transition?

Thank you!

**Comment by Heihokon** [comment]
Hi @drazenbuljovcic 

The fix is ​​released in version [3.2.1](https://www.npmjs.com/package/@flagship.io/js-sdk)

Thanks !
